docs/setup/settings.md
# API Settings

Keystone-API reads application settings from environment variables.
Individual settings are listed below by category and use case.

## Security Settings

Security settings are used to configure application networking and request signing.
These values should be chosen with care.
Improperly configured settings can introduce dangerous vulnerabilities and may damage your production deployment.

### Core Security

Keystone-API requires a random secret key to sign and verify requests.
Secret keys are conventionally 50 characters long and can be generated using common utilities like `openssl`.
For example: `openssl rand -base64 48 | cut -c1-50`

| Setting Name        | Default Value      | Description                                      |
|---------------------|--------------------|--------------------------------------------------|
| `SECURE_SECRET_KEY` | Randomly generated | Key value used to enforce cryptographic signing. |

### SSL/TLS

Enabling TLS is strongly recommended in production.
Enabling HSTS is also recommended, but only when TLS is already fully configured.
Administrators are cautioned to consider the potentially irreversible side effects of HSTS before enabling it.

Alternatively, administrators can implement TLS/HSTS support using a reverse proxy placed in front of the API.
This is a common approach in production environments and allows for centralized certificate management.
When using a reverse proxy for TLS, the settings below can be left at their default values.

| Setting Name             | Default Value  | Description                                       |
|--------------------------|----------------|---------------------------------------------------|
| `SECURE_SSL_REDIRECT`    | `False`        | Automatically redirect all HTTP traffic to HTTPS. |
| `SECURE_HSTS_SECONDS`    | `0` (Disabled) | HSTS cache duration in seconds.                   |
| `SECURE_HSTS_SUBDOMAINS` | `False`        | Enable HSTS for subdomains.                       |
| `SECURE_HSTS_PRELOAD`    | `False`        | Enable HSTS preload functionality.                |

### Auth Tokens

The following settings define which domains are allowed to interact with the Keystone-API.
In most deployments, these settings should be configured as follows:

- **`SECURE_SSL_TOKENS`**: Enabled when serving over HTTPS and disabled otherwise.
- **`SECURE_ALLOWED_HOSTS`**: Set to the domain name(s) where the API is hosted <br> (e.g., `api.example.com`).
- **`SECURE_ALLOWED_ORIGINS`**: Set to the full URL(s) of the frontend web application <br> (e.g., `https://app.example.com`).
- **`SECURE_TOKEN_DOMAIN`**: Only required when the API and frontend are hosted on different subdomains of the same
  parent domain. This setting should specify the parent domain with a leading dot (e.g., `.example.com`) to allow 
  token sharing across subdomains.

Default values are defined relative to the following list of _default local addresses_:

- `http://localhost:80`
- `https://localhost:443`
- `http://localhost:4200`
- `http://localhost:8000`
- `http://127.0.0.1:80`
- `https://127.0.0.1:443`
- `http://127.0.0.1:4200`
- `http://127.0.0.1:8000`

| Setting Name             | Default Value                        <br/><br/> | Description                                                                                      |
|--------------------------|-------------------------------------------------|--------------------------------------------------------------------------------------------------|
| `SECURE_SSL_TOKENS`      | `False`                                         | Only issue session/CSRF tokens over secure connections.                                          |
| `SECURE_ALLOWED_HOSTS`   | <code>localhost,127.0.0.1</code>                | Comma-separated list of api host/domain names (**without** protocol).                            |
| `SECURE_ALLOWED_ORIGINS` | _See default local addresses._                  | Comma-separated list of accepted client origin domains (**with** protocol).                      |
| `SECURE_SESSION_AGE`     | `1209600` (2 weeks)                             | Number of seconds before session tokens expire.                                                  |
| `SECURE_TOKEN_DOMAIN`    | None                                            | Domain attribute for session/csrf cookies. Set for cross-subdomain usage (e.g., `.example.com`). |

## General Configuration

Keystone uses various static files and user content to facilitate operation.
By default, these files are stored in subdirectories of the installed application directory (`<app>`).

| Setting Name           | Default Value         | Description                                                              |
|------------------------|-----------------------|--------------------------------------------------------------------------|
| `CONFIG_TIMEZONE`      | `UTC`                 | The timezone to use when rendering date/time values.                     |
| `CONFIG_STATIC_DIR`    | `<app>/static_files`  | Where to store internal static files required by the application.        |
| `CONFIG_UPLOAD_DIR`    | `<app>/media`         | Where to store file data uploaded by users.                              |
| `CONFIG_UPLOAD_SIZE`   | `2621440` (2.5 MB)    | Maximum allowed file upload size in bytes.                               |
| `CONFIG_UPLOAD_COUNT`  | `15`                  | Maximum files allowed in a single upload request.                        |
| `CONFIG_METRICS_PORTS` | `9101` through `9150` | Port numbers used to expose prometheus metrics (e.g., `9101,9102,9103`). |

## Logging

Keystone automatically purges log records according to the policy settings below.
Application logs are written to disk using a size-based policy that rotates files according to a maximum file size/count.
Audit, request, and task logs are maintained in the application database and are removed once they exceed a configured
age (in seconds).

| Setting Name              | Default Value        | Description                                                                                                 |
|---------------------------|----------------------|-------------------------------------------------------------------------------------------------------------|
| `LOG_APP_LEVEL`           | `WARNING`            | Only record application logs above this level (accepts `CRITICAL`, `ERROR`, `WARNING`, `INFO`, or `DEBUG`). |
| `LOG_APP_FILE`            | `<app>/keystone.log` | Destination file path for application logs.                                                                 |
| `LOG_APP_RETENTION_BYTES` | `10485760` (10 MB)   | Maximum log file size before rotating log files.                                                            |
| `LOG_APP_RETENTION_FILES` | `5`                  | Maximum rotated log files to keep.                                                                          |
| `LOG_REQ_RETENTION_SEC`   | `2592000` (30 days)  | How long to store request logs in seconds. Set to 0 to keep all records.                                    |
| `LOG_AUD_RETENTION_SEC`   | `2592000` (30 days)  | How long to store audit logs in seconds. Set to 0 to keep all records.                                      |
| `LOG_TSK_RETENTION_SEC`   | `2592000` (30 days)  | How long to store task logs in seconds. Set to 0 to keep all records.                                       |

## API Throttling

API settings are used to throttle incoming API requests against a maximum limit.
Limits are specified as the maximum number of requests per `day`, `minute`, `hour`, or `second`.

| Setting Name        | Default Value | Description                                          |
|---------------------|---------------|------------------------------------------------------|
| `API_THROTTLE_ANON` | `120/min`     | Rate limiting for anonymous (unauthenticated) users. |
| `API_THROTTLE_USER` | `300/min`     | Rate limiting for authenticated users.               |

## Database Connection

Official support is included for both SQLite and PostgreSQL database backends.
Using SQLite is intended for development and demonstrative use cases only.
The PostgreSQL backend should always be used in production settings.

| Setting Name         | Default Value | Description                                             |
|----------------------|---------------|---------------------------------------------------------|
| `DB_POSTGRES_ENABLE` | `False`       | Use PostgreSQL instead of the demo SQLite database.     |
| `DB_NAME`            | `keystone`    | The name of the application database.                   |
| `DB_USER`            |               | Username for database authentication (PostgreSQL only). |
| `DB_PASSWORD`        |               | Password for database authentication (PostgreSQL only). |
| `DB_HOST`            | `localhost`   | Database host address (PostgreSQL only).                |
| `DB_PORT`            | `5432`        | Database host port (PostgreSQL only).                   |

## Redis Connection

Redis settings define the network location and connection information for the application Redis cache.
Enabling password authentication is strongly recommended.

| Setting Name     | Default Value | Description                                  |
|------------------|---------------|----------------------------------------------|
| `REDIS_HOST`     | `127.0.0.1`   | Host address for the Redis message cache.    |
| `REDIS_PORT`     | `6379`        | Port number for the Redis message cache.     |
| `REDIS_DB`       | `0`           | The Redis database number to use.            |
| `REDIS_PASSWORD` |               | Optionally connect using the given password. |

## Email Notifications

Keystone will default to using the local server when issuing email notifications.
An alternative SMTP server can be specified using the settings below.
Securing your production email server with a username/password is strongly recommended.

| Setting Name          | Default Value             | Description                                             |
|-----------------------|---------------------------|---------------------------------------------------------|
| `EMAIL_HOST`          | `localhost`               | The host server to use for sending email.               |
| `EMAIL_HOST_USER`     |                           | Username to use for the SMTP server.                    |
| `EMAIL_HOST_PASSWORD` |                           | Password to use for the SMTP server.                    |
| `EMAIL_PORT`          | `25`                      | Port to use for the SMTP server.                        |
| `EMAIL_USE_TLS`       | `False`                   | Use a TLS connection to the SMTP server.                |
| `EMAIL_FROM_ADDRESS`  | `noreply@keystone.bot`    | The default "from" address used in email notifications. |
| `EMAIL_TEMPLATE_DIR`  | `/etc/keystone/templates` | Directory to search for customized email templates.     |
| `EMAIL_DEBUG_DIR`     |                           | Write emails to disk instead of using the SMTP server.  |

## LDAP Authentication

Using LDAP for authentication is optional and disabled by default.
To enable LDAP, set the `AUTH_LDAP_SERVER_URI` value to the desired LDAP endpoint.
Enabling LDAP integration will also add LDAP related health checks to the
[API health endpoint](../api/logging.md#system-health).

Application user fields are mapped to LDAP attributes by specifying the `AUTH_LDAP_ATTR_MAP` setting.
The following example maps the `first_name` and `last_name` fields used by Keystone to the LDAP attributes `givenName`
and `sn`:

```bash
AUTH_LDAP_ATTR_MAP="first_name=givenName,last_name=sn"
```

A full list of available Keystone fields can be found in the project's [OpenApi specification](../api/openapi.md).

| Setting Name              | Default Value           | Description                                                       |
|---------------------------|-------------------------|-------------------------------------------------------------------|
| `AUTH_LDAP_SERVER_URI`    |                         | The URI of the LDAP server.                                       |
| `AUTH_LDAP_START_TLS`     | `True`                  | Whether to use TLS when connecting to the LDAP server.            |
| `AUTH_LDAP_BIND_DN`       |                         | Optionally bind LDAP queries to the given DN.                     |
| `AUTH_LDAP_BIND_PASSWORD` |                         | The password to use when binding to the LDAP server.              |
| `AUTH_LDAP_USER_SEARCH`   |                         | The base DN for searching users in the LDAP server.               |
| `AUTH_LDAP_USER_FILTER`   | `(objectClass=account)` | Search filter to apply when selecting LDAP user account entries.  |
| `AUTH_LDAP_LOGIN_FILTER`  | `(uid=%(user)s)`        | Search filter for username lookups during authentication.         |
| `AUTH_LDAP_REQUIRE_CERT`  | `False`                 | Whether to require certificate verification.                      |
| `AUTH_LDAP_ATTR_MAP`      |                         | A mapping of user fields to LDAP attribute names.                 |
| `AUTH_LDAP_TIMEOUT`       | `10`                    | The number of seconds before timing out an LDAP connection/query. |

keystone_api/main/settings.py
"""Top level Django application settings."""

import importlib.metadata
import os
import sys
from datetime import timedelta
from pathlib import Path

import environ
from django.core.management.utils import get_random_secret_key
from jinja2 import StrictUndefined

BASE_DIR = Path(__file__).resolve().parent.parent
sys.path.insert(0, str(BASE_DIR))

# Application metadata

dist = importlib.metadata.distribution("keystone-api")
VERSION = dist.metadata["version"]
SUMMARY = dist.metadata["summary"]

env = environ.Env()

# Core security settings

_trusted_local = [
    "http://localhost:80",
    "https://localhost:443",
    "http://localhost:4200",
    "http://localhost:8000",
    "http://127.0.0.1:80",
    "https://127.0.0.1:443",
    "http://127.0.0.1:4200",
    "http://127.0.0.1:8000",
]

SECRET_KEY = os.environ.get("SECURE_SECRET_KEY", get_random_secret_key())
ALLOWED_HOSTS = env.list("SECURE_ALLOWED_HOSTS", default=["localhost", "127.0.0.1"])

_SECURE_SSL_TOKENS = env.bool("SECURE_SSL_TOKENS", False)
SESSION_COOKIE_SECURE = _SECURE_SSL_TOKENS
SESSION_COOKIE_HTTPONLY = True
SESSION_COOKIE_SAMESITE = "Lax"
SESSION_COOKIE_AGE = env.int("SECURE_SESSION_AGE", timedelta(days=14).total_seconds())

SECURE_SSL_REDIRECT = env.bool("SECURE_SSL_REDIRECT", False)
SECURE_HSTS_PRELOAD = env.bool("SECURE_HSTS_PRELOAD", False)
SECURE_HSTS_SECONDS = env.int("SECURE_HSTS_SECONDS", 0)
SECURE_HSTS_INCLUDE_SUBDOMAINS = env.bool("SECURE_HSTS_SUBDOMAINS", False)

_ALLOWED_ORIGINS = env.list("SECURE_ALLOWED_ORIGINS", default=_trusted_local)
CORS_ALLOW_CREDENTIALS = True
CORS_ALLOWED_ORIGINS = _ALLOWED_ORIGINS
CORS_ALLOW_HEADERS = [
    "accept",
    "authorization",
    "content-type",
    "user-agent",
    "x-csrftoken",
    "x-requested-with",
    "x-keystone-cid",
]

CSRF_TRUSTED_ORIGINS = _ALLOWED_ORIGINS
CSRF_COOKIE_SECURE = _SECURE_SSL_TOKENS
CSRF_COOKIE_HTTPONLY = False
CSRF_COOKIE_SAMESITE = "Lax"

_cookie_domain = os.environ.get("SECURE_TOKEN_DOMAIN", None)
SESSION_COOKIE_DOMAIN = _cookie_domain
CSRF_COOKIE_DOMAIN = _cookie_domain

ALLOWED_FILE_TYPES = [
    # Documents
    "application/pdf",
    "application/rtf",
    "application/msword",  # .doc
    "application/vnd.openxmlformats-officedocument.wordprocessingml.document",  # .docx
    "application/vnd.oasis.opendocument.text",  # .odt
    "application/x-latex",  # .latex
    "application/x-tex",  # .tex

    # Spreadsheets
    "application/vnd.ms-excel",  # .xls
    "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",  # .xlsx
    "application/vnd.oasis.opendocument.spreadsheet",  # .ods

    # Presentations
    "application/vnd.ms-powerpoint",  # .ppt
    "application/vnd.openxmlformats-officedocument.presentationml.presentation",  # .pptx

    # Plain and structured text
    "text/plain",  # .txt
    "text/markdown",  # .md
    "text/richtext",  # .rtx
    "text/csv",  # .csv

    # Images
    "image/png",
    "image/jpeg",
    "image/gif",
    "image/svg+xml",
    "image/tiff",
    "image/bmp"
]

# Trust system administrators and disable select deployment checks
SILENCED_SYSTEM_CHECKS = [
    "security.W004",  # SECURE_HSTS_SECONDS is not set.
    "security.W008",  # SECURE_SSL_REDIRECT setting is not set to True.
    "security.W012",  # SESSION_COOKIE_SECURE is not set to True.
    "security.W016"  # CSRF_COOKIE_SECURE is not set to True.
]

# App Configuration

ROOT_URLCONF = "main.urls"
LOGIN_REDIRECT_URL = "/"
SITE_ID = 1

INSTALLED_APPS = [
    "jazzmin",
    "auditlog",
    "corsheaders",
    "django.contrib.admin",
    "django.contrib.auth",
    "django.contrib.contenttypes",
    "django.contrib.sessions",
    "django.contrib.messages",
    "django.contrib.staticfiles",
    "django.contrib.sites",
    "health_check",
    "rest_framework",
    "rest_framework.authtoken",
    "django_celery_beat",
    "django_celery_results",
    "django_filters",
    "django_prometheus",
    "drf_spectacular",
    "plugins",
    "apps.admin_utils",
    "apps.allocations",
    "apps.authentication",
    "apps.batch",
    "apps.config",
    "apps.health",
    "apps.logging",
    "apps.notifications",
    "apps.openapi",
    "apps.research_products",
    "apps.scheduler",
    "apps.stats",
    "apps.users",
]

MIDDLEWARE = [
    "django_prometheus.middleware.PrometheusBeforeMiddleware",
    "corsheaders.middleware.CorsMiddleware",
    "apps.logging.middleware.LogRequestMiddleware",
    "django.middleware.security.SecurityMiddleware",
    "django.contrib.sessions.middleware.SessionMiddleware",
    "django.middleware.common.CommonMiddleware",
    "django.middleware.csrf.CsrfViewMiddleware",
    "django.contrib.auth.middleware.AuthenticationMiddleware",
    "django.contrib.messages.middleware.MessageMiddleware",
    "django.middleware.clickjacking.XFrameOptionsMiddleware",
    "servestatic.middleware.ServeStaticMiddleware",
    "auditlog.middleware.AuditlogMiddleware",
    "django_prometheus.middleware.PrometheusAfterMiddleware",
]

TEMPLATES = [
    {  # The default backend required by Django builtins (e.g., the admin)
        "BACKEND": "django.template.backends.django.DjangoTemplates",
        "APP_DIRS": True,
        "OPTIONS": {
            "context_processors": [
                "django.template.context_processors.debug",
                "django.template.context_processors.request",
                "django.contrib.auth.context_processors.auth",
                "django.contrib.messages.context_processors.messages",
            ],
        },
    },
]

# Base styling for the Admin UI

USE_THOUSAND_SEPARATOR = True
JAZZMIN_SETTINGS = {
    "site_title": "Keystone",
    "site_header": "Keystone",
    "site_brand": "Keystone",
    "hide_apps": ["sites", "auth", "authtoken", "token_blacklist", "django_celery_beat"],
    "order_with_respect_to": [
        "users",
        "allocations",
        "research_products",
        "sites"
    ],
    "icons": {},
    "login_logo": "fake/file/path.jpg",  # Missing file path hides the logo
}

# REST API settings

REST_FRAMEWORK = {
    "SEARCH_PARAM": "_search",
    "ORDERING_PARAM": "_order",
    "DEFAULT_AUTHENTICATION_CLASSES": [
        "rest_framework.authentication.TokenAuthentication",
        "rest_framework.authentication.SessionAuthentication",
    ],
    "DEFAULT_PERMISSION_CLASSES": [
        "rest_framework.permissions.IsAuthenticated"
    ],
    "DEFAULT_THROTTLE_CLASSES": [
        "rest_framework.throttling.AnonRateThrottle",
        "rest_framework.throttling.UserRateThrottle"
    ],
    "DEFAULT_THROTTLE_RATES": {
        "anon": env.str("API_THROTTLE_ANON", "120/min"),
        "user": env.str("API_THROTTLE_USER", "300/min")
    },
    "DEFAULT_RENDERER_CLASSES": [
        "rest_framework.renderers.JSONRenderer",
    ],
    "DEFAULT_FILTER_BACKENDS": (
        "plugins.filters.AutoFilterBackend",
        "rest_framework.filters.OrderingFilter",
        "rest_framework.filters.SearchFilter"
    ),
    "DEFAULT_PAGINATION_CLASS": "plugins.pagination.PaginationHandler",
    "DEFAULT_SCHEMA_CLASS": "drf_spectacular.openapi.AutoSchema",
}

# Audit logging

AUDITLOG_CID_HEADER = "X-KEYSTONE-CID"  # Use uppercase and dashes

# Customize the generation of OpenAPI specifications

SPECTACULAR_SETTINGS = {
    "TITLE": "Keystone API",
    "DESCRIPTION": SUMMARY,
    "VERSION": VERSION,
    "SERVE_INCLUDE_SCHEMA": False,
    "COMPONENT_SPLIT_REQUEST": True,
    "COMPONENT_SPLIT_PATCH": True,
    "ENUM_NAME_OVERRIDES": {
        "RequestStatusChoices": "apps.allocations.models.AllocationRequest.StatusChoices",
        "ReviewStatusChoices": "apps.allocations.models.AllocationReview.StatusChoices",
    },
    "POSTPROCESSING_HOOKS": [
        "plugins.schemas.mark_all_get_fields_required",
    ],
}

# Redis backend and Celery scheduler

_redis_host = env.url("REDIS_HOST", "127.0.0.1").geturl()
_redis_port = env.int("REDIS_PORT", 6379)
_redis_db = env.int("REDIS_DB", 0)
_redis_pass = env.str("REDIS_PASSWORD", "")

REDIS_URL = f"redis://:{_redis_pass}@{_redis_host}:{_redis_port}"

CELERY_BROKER_URL = REDIS_URL + f"/{_redis_db}"
CELERY_CACHE_BACKEND = "django-cache"
CELERY_RESULT_BACKEND = "django-db"
CELERY_RESULT_EXTENDED = True

# Email server

EMAIL_FROM_ADDRESS = env.str("EMAIL_FROM_ADDRESS", "noreply@keystone.bot")
EMAIL_TEMPLATE_DIR = env.path("EMAIL_TEMPLATE_DIR", Path("/etc/keystone/templates"))
EMAIL_DEFAULT_DIR = BASE_DIR / "templates"

if _email_path := env.get_value("EMAIL_DEBUG_DIR", default=None):
    EMAIL_BACKEND = "django.core.mail.backends.filebased.EmailBackend"
    EMAIL_FILE_PATH = _email_path

else:
    EMAIL_BACKEND = "django.core.mail.backends.smtp.EmailBackend"
    EMAIL_HOST = env.str("EMAIL_HOST", "localhost")
    EMAIL_PORT = env.int("EMAIL_PORT", 25)
    EMAIL_HOST_USER = env.str("EMAIL_HOST_USER", "")
    EMAIL_HOST_PASSWORD = env.str("EMAIL_HOST_PASSWORD", "")
    EMAIL_USE_TLS = env.bool("EMAIL_USE_TLS", False)

# Database

DATABASES = dict()
DEFAULT_AUTO_FIELD = "django.db.models.BigAutoField"

_db_name = env.str("DB_NAME", "keystone")
_db_path = (BASE_DIR / _db_name).with_suffix(".db")
if env.bool("DB_POSTGRES_ENABLE", False):
    DATABASES["default"] = {
        "ENGINE": "django_prometheus.db.backends.postgresql",
        "NAME": _db_name,
        "USER": env.str("DB_USER", ""),
        "PASSWORD": env.str("DB_PASSWORD", ""),
        "HOST": env.str("DB_HOST", "localhost"),
        "PORT": env.str("DB_PORT", "5432"),
    }

else:
    DATABASES["default"] = {
        "ENGINE": "django.db.backends.sqlite3",
        "NAME": _db_path,
        "timeout": 30,
        "PRAGMA": {
            "journal_mode": "wal",
        }
    }

# Authentication

AUTH_USER_MODEL = "users.User"
AUTHENTICATION_BACKENDS = ["django.contrib.auth.backends.ModelBackend"]

# General LDAP settings
AUTH_LDAP_TIMEOUT = env.int("AUTH_LDAP_TIMEOUT", 10)
AUTH_LDAP_ALWAYS_UPDATE_USER = True
AUTH_LDAP_START_TLS = env.bool("AUTH_LDAP_START_TLS", True)
AUTH_LDAP_BIND_DN = env.str("AUTH_LDAP_BIND_DN", "")
AUTH_LDAP_BIND_PASSWORD = env.str("AUTH_LDAP_BIND_PASSWORD", "")
AUTH_LDAP_USER_ATTR_MAP = env.dict("AUTH_LDAP_ATTR_MAP", default=dict())
AUTH_LDAP_USER_FILTER = env.str("AUTH_LDAP_USER_FILTER", "(objectClass=account)")

# LDAP settings that require the LDAP dependency to initialize
if AUTH_LDAP_SERVER_URI := env.url("AUTH_LDAP_SERVER_URI", "").geturl():
    import ldap
    from django_auth_ldap.config import LDAPSearch

    AUTHENTICATION_BACKENDS.append("django_auth_ldap.backend.LDAPBackend")
    AUTH_LDAP_USER_SEARCH = LDAPSearch(
        env.str("AUTH_LDAP_USER_SEARCH", ""),
        ldap.SCOPE_SUBTREE,
        env.str("AUTH_LDAP_LOGIN_FILTER", "(uid=%(user)s)")
    )

    if env.bool("AUTH_LDAP_REQUIRE_CERT", False):
        AUTH_LDAP_GLOBAL_OPTIONS = {ldap.OPT_X_TLS_REQUIRE_CERT: ldap.OPT_X_TLS_NEVER}

    else:
        AUTH_LDAP_GLOBAL_OPTIONS = {ldap.OPT_X_TLS_REQUIRE_CERT: ldap.OPT_X_TLS_TRY}

AUTH_PASSWORD_VALIDATORS = [
    {"NAME": "django.contrib.auth.password_validation.UserAttributeSimilarityValidator"},
    {"NAME": "django.contrib.auth.password_validation.MinimumLengthValidator"},
    {"NAME": "django.contrib.auth.password_validation.CommonPasswordValidator"},
    {"NAME": "django.contrib.auth.password_validation.NumericPasswordValidator"},
]

# Static file handling (CSS, JavaScript, Images)

MAX_FILE_SIZE = env.int("CONFIG_UPLOAD_SIZE", 2.5 * 1024 * 1024)  # 2.5 MB
MAX_FILE_COUNT = env.int("CONFIG_UPLOAD_COUNT", 15)

STATIC_URL = "/static/"
STATIC_ROOT = Path(env.path("CONFIG_STATIC_DIR", BASE_DIR / "static_files"))
STATIC_ROOT.mkdir(mode=0o770, parents=True, exist_ok=True)

MEDIA_URL = "/media/"
MEDIA_ROOT = Path(env.path("CONFIG_UPLOAD_DIR", BASE_DIR / "media"))
MEDIA_ROOT.mkdir(mode=0o770, parents=True, exist_ok=True)

# Timezones

USE_TZ = True
CELERY_ENABLE_UTC = True
DJANGO_CELERY_BEAT_TZ_AWARE = True
TIME_ZONE = env.str("CONFIG_TIMEZONE", "UTC")

# Prometheus Metrics

PROMETHEUS_METRICS_EXPORT_PORT_RANGE = env.list("CONFIG_METRICS_PORTS", default=range(9101, 9150), cast=int)

# Logging

# Disable Celery's internal log clean up in favor of custom log rotation
CELERY_RESULT_EXPIRES = None

LOG_REQ_RETENTION_SEC = env.int("LOG_REQ_RETENTION_SEC", timedelta(days=30).total_seconds())
LOG_AUD_RETENTION_SEC = env.int("LOG_AUD_RETENTION_SEC", timedelta(days=30).total_seconds())
LOG_TSK_RETENTION_SEC = env.int("LOG_TSK_RETENTION_SEC", timedelta(days=30).total_seconds())

_default_log_dir = BASE_DIR / "keystone.log"
LOG_FILE_PATH = Path(os.getenv("LOG_APP_FILE", _default_log_dir))
LOG_FILE_PATH.parent.mkdir(mode=0o770, parents=True, exist_ok=True)

LOGGING = {
    "version": 1,
    "disable_existing_loggers": False,
    "formatters": {
        "verbose": {
            "format": "%(asctime)s [%(levelname)s] %(name)s: %(message)s"
        },
    },
    "handlers": {
        "file": {
            "level": env.str("LOG_APP_LEVEL", "WARNING"),
            "class": "logging.handlers.RotatingFileHandler",
            "filename": str(LOG_FILE_PATH),
            "maxBytes": env.int("LOG_APP_RETENTION_BYTES", 10 * 1024 * 1024),  # Default 10 MB
            "backupCount": env.int("LOG_APP_RETENTION_FILES", 5),  # Default 5 backups
            "formatter": "verbose",
        },
    },
    "loggers": {
        "": {
            "handlers": ["file"],
            "level": env.str("LOG_APP_LEVEL", "WARNING"),
        },
        "apps": {
            "handlers": ["file"],
            "level": env.str("LOG_APP_LEVEL", "WARNING"),
            "propagate": False,
        },
    }
}
